istioctl/cmd/istiodconfig.go

```diff
@@ -0,0 +1,464 @@
+// Copyright Istio Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmd
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+	"net/url"
+	"regexp"
+	"sort"
+	"strings"
+	"sync"
+
+	"github.com/spf13/cobra"
+
+	"istio.io/istio/istioctl/pkg/clioptions"
+	"istio.io/istio/istioctl/pkg/util/handlers"
+	"istio.io/pkg/log"
+)
+
+type flagState interface {
+	run() (string, error)
+}
+
+var _ flagState = (*resetState)(nil)
+var _ flagState = (*logLevelState)(nil)
+var _ flagState = (*stackTraceLevelState)(nil)
+var _ flagState = (*getAllLogLevelsState)(nil)
+
+type resetState struct {
+	client *ControlzClient
+}
+
+func (rs *resetState) run() (string, error) {
+
+	const (
+		defaultOutputLevel     = "info"
+		defaultStackTraceLevel = "none"
+	)
+	allScopes, err := rs.client.GetScopes()
+	if err != nil {
+		return "", fmt.Errorf("could not get all scopes: %v", err)
+	}
+	var ch = make(chan struct {
+		err       error
+		scopeName string
+	}, len(allScopes))
+	var wg sync.WaitGroup
+	for _, scope := range allScopes {
+		wg.Add(1)
+		go func(scope *ScopeInfo) {
+			defer wg.Done()
+			err := rs.client.PutScope(&ScopeInfo{
+				Name:            scope.Name,
+				OutputLevel:     defaultOutputLevel,
+				StackTraceLevel: defaultStackTraceLevel})
+			ch <- struct {
+				err       error
+				scopeName string
+			}{err: err, scopeName: scope.Name}
+		}(scope)
+	}
+	wg.Wait()
+	close(ch)
+	for result := range ch {
+		if result.err != nil {
+			return "", fmt.Errorf("failed updating scope %s: %v", result.scopeName, result.err)
+
+		}
+	}
+	return "", nil
+}
+
+type logLevelState struct {
+	client            *ControlzClient
+	outputLogLevel    string
+	validationPattern string
+}
+
+func (ll *logLevelState) run() (string, error) {
```

Comment:
Seems like this and reset and stackTraceLevelState should all be calling the same function and just passing in what to set. That way code is re-used - currently its not only repeated but also different; for some reason reset is parallel but this is not?

Reply:
Thanks for feedback. I've refactored this a bit moving api calls concurrent to controlz Client